### DISCUSSION TITLE
Feature Request: Show Changelogs for Elixir / Mix / Hex

### DISCUSSION THREAD

**Discussion by skyqrose** [renovatebot/renovate#29993] [Suggest an Idea]
### Tell us more.

Updates for Elixir dependencies don't include changelogs in the PR description, so I have to go chase down a changelog for every update, which is tedious. Changelogs do work for npm updates in the same repo.

Here's an example PR without a changelog: https://togithub.com/sheerlox/elixir_renovate_demo/pull/25 (note: it's not my repo, just a random public one I found, cuz I only have renovate installed on a private repo.)

Note that this PR for a Mix dependency for a library hosted on GitHub instead of Hex _does_ have an included changelog: https://togithub.com/sheerlox/elixir_renovate_demo/pull/8 (again, not my repo).

The renovate PRs do include links to the hex.pm page. Example of an included link: https://hex.pm/packages/credo

There are 4 places I typically access the changelog via that page. (This `credo` example has all 4, but other packages may be missing any or all of them.)
- Direct link to changelog from that page (in this case, links to the GitHub CHANGELOG.md)
- GitHub CHANGELOG.md file: https://togithub.com/rrrene/credo/blob/master/CHANGELOG.md (via link to GitHub)
- GitHub Releases: https://togithub.com/rrrene/credo/releases (via link to GitHub)
- Link to "Online Documentation" (hexdocs.pm), and then find the changelog in the left navigation: https://hexdocs.pm/credo/changelog.html

Here's all the docs I did find about this:
- Issue for tracking a bunch of changelog fetching improvements: https://togithub.com/renovatebot/renovate/issues/14138
- Docs: https://docs.renovatebot.com/key-concepts/changelogs/#platforms-that-renovate-can-fetch-changelogs-from
  - I see that there's a limited number of supported platforms. GitHub is listed, but Hex isn't. The changelog is hosted on GitHub, but is the problem that Renovate can't find it because it has to go through Hex?
  - The troubleshooting step 1 is "The datasource for this package does not support sourceUrls." but I don't know how I'd go about finding out whether that's the case.
- (I looked and didn't see any existing issues or discussion on Elixir changelogs.)

Is this an easy configuration fix for the Mix manager or Hex datasource, or does it depend on upgrades to the renovate core to handle other platforms? (Or is it just a problem with my repo's configuration.)

Relevant labels: `datasource:hex`, `manager:mix`, `core:changelogs`

**Comment by rarkins** [comment]
If changelogs work for public, but not for private, then please rewrite/clarify the above to indicate that your request is about private changelogs specifically.

If changelogs don't work at all, for private or public, then please de-emphasize the private part of your description and find a way to reproduce what you are asking for using public packages/repos. e.g. an update PR from a public package/repository which is missing changelogs.

**Reply by skyqrose** [reply]
I currently only use renovate on a single private repo, so I've updated the description with a link to a somebody's public repo that shows the same problem.

Also, to clarify, my repo that renovate is installed in is private, but the dependencies and changelogs it's looking up are public.

**Comment by rarkins** [comment]
Here is an example URL which is queried: https://hex.pm/api/packages/phoenix_ecto

It has a meta.links.GitHub URL

If I look at the test case we have in the Renovate repo, it's for https://hex.pm/api/packages/certifi

It has a meta.links.Github URL (note the lower-case `h`).

I guess both are ok.

@zharinov do you know how to make a key in zod case-insensitive?

**Reply by bernardo-martinez** [reply]
hello, this would be super helpful for us.
If it is not able to handle case-insensitive keys, what about pre normalising them with an utility function:

```
function normalizeKeys(obj: Record<string, any>): Record<string, any> {
  return Object.fromEntries(
    Object.entries(obj).map(([key, value]) => [key.toLowerCase(), value])
  );
}
```

and then:
```
...
  const normalizedLinks = normalizeKeys(hexResponse.meta?.links);
  if (normalizedLinks?.github)
      releaseResult.sourceUrl = normalizedLinks.github;
...
```

**Reply by bernardo-martinez** [reply]
wdyt @rarkins ?

**Reply by rarkins** [reply]
PR would be welcome if this approach works

**Reply by bernardo-martinez** [reply]
thanks, PR in my fork made [here](https://github.com/renovatebot/renovate/pull/32963)
